Implement Packge::from_file()

dralley opened the pull request:
It makes more sense for Package to know details about Header than the other way around.  And we probably want some permutation of a `Package::from_file()` function to exist which will involve using functions that deal with headers.

edit: 5 years later

additional changes:

Package no longer copies all tag values into owned String fields up
front. Instead it holds a reference-counted Header and performs tag
lookups on demand through the existing accessors.

Header gains a Clone impl that increments librpm's internal reference
count via headerLink(), matching the existing Drop impl that calls
headerFree().

Package's derived trait impls (Clone, Debug, PartialEq, Eq, Hash) are
replaced with manual implementations that delegate to tag lookups.
PartialEq and Hash compare by NEVRA components (name, epoch, version,
release, arch).

New public API:
- Package::from_file() reads an .rpm file directly into a Package
- Package::get() exposes raw tag data access via Tag and TagData,
  both of which are now part of the public API
- RpmErrorKind is re-exported for from_file() error handling

tarcieri commented on src/package.rs:
```diff
@@ -1,35 +1,50 @@
 }
 
 impl Package {
+    pub(crate) fn from_header(h: &Header) -> Self {
```
Since we're taking a look at this one again, I think it might make sense to make this a `Result<Self, ...>` and propagating the errors below, rather than using `unwrap()`

dralley commented on src/internal/header.rs:
```diff
@@ -1,20 +1,66 @@
+
+        let fd: librpm_sys::FD_t = unsafe { librpm_sys::Fopen(filename.as_ptr(), fmode.as_ptr()) };
+        // GlobalTS.lock() ?  Or can the transaction sets be independent because we're not touching a global database?
+        let ts = TransactionSet::create();
```
@tarcieri This PR isn't ready for review (it doesn't even compile) but I have a question about how this works.

tarcieri replied:
Yeah, I should probably write up some documentation on how those work, at least in the current form.

It's one of those things that went through multiple rewrites trying different strategies after long, protracted discussions with RPM maintainers, and even then there are still bugs in the current implementation I believe.

tarcieri replied:
To perhaps more directly answer your question, the current locking model uses `GlobalTS` for everything, in all cases.

dralley committed:
Turn Package into a thin wrapper around Header

Package no longer copies all tag values into owned String fields up
front. Instead it holds a reference-counted Header and performs tag
lookups on demand through the existing accessors.

Header gains a Clone impl that increments librpm's internal reference
count via headerLink(), matching the existing Drop impl that calls
headerFree().

Package's derived trait impls (Clone, Debug, PartialEq, Eq, Hash) are
replaced with manual implementations that delegate to tag lookups.
PartialEq and Hash compare by NEVRA components (name, epoch, version,
release, arch).

New public API:
- Package::from_file() reads an .rpm file directly into a Package
- Package::get() exposes raw tag data access via Tag and TagData,
  both of which are now part of the public API
- RpmErrorKind is re-exported for from_file() error handling

Assisted-By: Claude Opus 4.6
